
## Extraccion, Tranformacion y Carga

In [1]:
import json             # importo modulo json encoder and decoder
import pandas as pd     # importo modulo panda Data Analysis Library
import ast              # importo modulo Abstract Syntax Trees

Una vez importados los modulos, importo el dataset e inicio mi trabajo con pandas

      Incorporo el archivo australian_user_items

In [2]:
reviews = []

with open('Data/australian_users_items.json', 'r', encoding='utf-8') as f:
    # utilizo un bucle for para insertar datos en la lista
    for line in f.readlines():
        reviews.append(ast.literal_eval(line))
# creo el dataframe con los datos de la lista reviews
df_user_items = pd.DataFrame(reviews)
# muestro el dataframe
df_user_items

user_id  items_count           steam_id  \
0            76561197970982479          277  76561197970982479   
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
88305        76561198323066619           22  76561198323066619   
88306        76561198326700687          177  76561198326700687   
88307  XxLaughingJackClown77xX            0  76561198328759259   
88308        76561198329548331            7  76561198329548331   
88309        edward_tremethick            0  76561198331598578   

                                                user_url  \
0      http://steamcommunity.com/profiles/76561197970...   
1                   http://steamcommunity.com/id/js41637   
2                 http://steamcommunity.com/id/evcentric   
3                http://steamcommunity.com/id/Riot-Punch   
4                     http://steamcommunity.com/id/doctr   
...                                                  ...   
88305  http://steamcommunity.com/profiles/76561198323...   
88306  http://steamcommunity.com/profiles/76561198326...   
88307  http://steamcommunity.com/id/XxLaughingJackClo...   
88308  http://steamcommunity.com/profiles/76561198329...   
88309     http://steamcommunity.com/id/edward_tremethick   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 5 columns]

Muestro informacion del dataFrame

In [3]:
df_user_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88310 entries, 0 to 88309
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      88310 non-null  object
 1   items_count  88310 non-null  int64 
 2   steam_id     88310 non-null  object
 3   user_url     88310 non-null  object
 4   items        88310 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.4+ MB


Verifico el contenido de la columna 'Items'

In [4]:
print(df_user_items['items'][0])

[{'item_id': '10', 'item_name': 'Counter-Strike', 'playtime_forever': 6, 'playtime_2weeks': 0}, {'item_id': '20', 'item_name': 'Team Fortress Classic', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '30', 'item_name': 'Day of Defeat', 'playtime_forever': 7, 'playtime_2weeks': 0}, {'item_id': '40', 'item_name': 'Deathmatch Classic', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '50', 'item_name': 'Half-Life: Opposing Force', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '60', 'item_name': 'Ricochet', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '70', 'item_name': 'Half-Life', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '130', 'item_name': 'Half-Life: Blue Shift', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '300', 'item_name': 'Day of Defeat: Source', 'playtime_forever': 4733, 'playtime_2weeks': 0}, {'item_id': '240', 'item_name': 'Counter-Strike: Source', 'playtime_forever': 1853, 'playtime_2weeks': 0}, {'item_i

La columna 'Items' tiene una lista con mas de un diccionario, los cuales contienen items que consumen los usuarios. Dentro de los diccionarios se distinguen las siguientes claves:

    Info , check Json keys

        {'item_id': '10',

        'item_name': 'Counter-Strike',

        'playtime_forever': 6,

        'playtime_2weeks': 0}

La columna 'Items' se encuentra anidada, por lo que se normaliza esta columna para obtener una clumna por cada clave encontrada en los diccionarios:

In [5]:
df_user_items= df_user_items.explode(["items"])
df_user_items = pd.json_normalize(reviews , record_path=["items"] , meta=['steam_id','items_count','user_id', 'user_url'] )
df_user_items

item_id                      item_name  playtime_forever  \
0            10                 Counter-Strike                 6   
1            20          Team Fortress Classic                 0   
2            30                  Day of Defeat                 7   
3            40             Deathmatch Classic                 0   
4            50      Half-Life: Opposing Force                 0   
...         ...                            ...               ...   
5153204  346330                   BrainBread 2                 0   
5153205  373330                    All Is Dust                 0   
5153206  388490  One Way To Die: Steam Edition                 3   
5153207  521570          You Have 10 Seconds 2                 4   
5153208  519140                     Minds Eyes                 3   

         playtime_2weeks           steam_id items_count            user_id  \
0                      0  76561197970982479         277  76561197970982479   
1                      0  76561197970982479         277  76561197970982479   
2                      0  76561197970982479         277  76561197970982479   
3                      0  76561197970982479         277  76561197970982479   
4                      0  76561197970982479         277  76561197970982479   
...                  ...                ...         ...                ...   
5153204                0  76561198329548331           7  76561198329548331   
5153205                0  76561198329548331           7  76561198329548331   
5153206                3  76561198329548331           7  76561198329548331   
5153207                4  76561198329548331           7  76561198329548331   
5153208                3  76561198329548331           7  76561198329548331   

                                                  user_url  
0        http://steamcommunity.com/profiles/76561197970...  
1        http://steamcommunity.com/profiles/76561197970...  
2        http://steamcommunity.com/profiles/76561197970...  
3        http://steamcommunity.com/profiles/76561197970...  
4        http://steamcommunity.com/profiles/76561197970...  
...                                                    ...  
5153204  http://steamcommunity.com/profiles/76561198329...  
5153205  http://steamcommunity.com/profiles/76561198329...  
5153206  http://steamcommunity.com/profiles/76561198329...  
5153207  http://steamcommunity.com/profiles/76561198329...  
5153208  http://steamcommunity.com/profiles/76561198329...  

[5153209 rows x 8 columns]

Muestro nuevamente informacion del dataframe resultante

In [6]:
df_user_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5153209 entries, 0 to 5153208
Data columns (total 8 columns):
 #   Column            Dtype 
---  ------            ----- 
 0   item_id           object
 1   item_name         object
 2   playtime_forever  int64 
 3   playtime_2weeks   int64 
 4   steam_id          object
 5   items_count       object
 6   user_id           object
 7   user_url          object
dtypes: int64(2), object(6)
memory usage: 314.5+ MB


Elimino la columnas 'playtime_2weeks', que resulta innecesaria para el proyecto

In [7]:
df_user_items.drop(['playtime_2weeks'], axis=1, inplace=True)

Verifico si existen duplicados

In [8]:
df_duplicate = df_user_items.loc[df_user_items.duplicated()]
df_duplicate

item_id                         item_name  playtime_forever  \
164294       20             Team Fortress Classic                 5   
164295       50         Half-Life: Opposing Force                 0   
164296       70                         Half-Life                 0   
164297      130             Half-Life: Blue Shift                 0   
164298      220                       Half-Life 2               198   
...         ...                               ...               ...   
4898223  213670  South Park™: The Stick of Truth™               725   
4898224  221910               The Stanley Parable                53   
4898225  261030      The Walking Dead: Season Two               253   
4898226  273110     Counter-Strike Nexon: Zombies                 0   
4898227     730  Counter-Strike: Global Offensive                 0   

                  steam_id items_count            user_id  \
164294   76561198084006094         109             Nikiad   
164295   76561198084006094         109             Nikiad   
164296   76561198084006094         109             Nikiad   
164297   76561198084006094         109             Nikiad   
164298   76561198084006094         109             Nikiad   
...                    ...         ...                ...   
4898223  76561198080057659          39  76561198080057659   
4898224  76561198080057659          39  76561198080057659   
4898225  76561198080057659          39  76561198080057659   
4898226  76561198080057659          39  76561198080057659   
4898227  76561198080057659          39  76561198080057659   

                                                  user_url  
164294                 http://steamcommunity.com/id/Nikiad  
164295                 http://steamcommunity.com/id/Nikiad  
164296                 http://steamcommunity.com/id/Nikiad  
164297                 http://steamcommunity.com/id/Nikiad  
164298                 http://steamcommunity.com/id/Nikiad  
...                                                    ...  
4898223  http://steamcommunity.com/profiles/76561198080...  
4898224  http://steamcommunity.com/profiles/76561198080...  
4898225  http://steamcommunity.com/profiles/76561198080...  
4898226  http://steamcommunity.com/profiles/76561198080...  
4898227  http://steamcommunity.com/profiles/76561198080...  

[59117 rows x 7 columns]

Se encontraron en total 59117 duplicados, por lo que se procede a elmininarlos

In [9]:
df_user_items = df_user_items.drop_duplicates(keep='first')
df_user_items

item_id                      item_name  playtime_forever  \
0            10                 Counter-Strike                 6   
1            20          Team Fortress Classic                 0   
2            30                  Day of Defeat                 7   
3            40             Deathmatch Classic                 0   
4            50      Half-Life: Opposing Force                 0   
...         ...                            ...               ...   
5153204  346330                   BrainBread 2                 0   
5153205  373330                    All Is Dust                 0   
5153206  388490  One Way To Die: Steam Edition                 3   
5153207  521570          You Have 10 Seconds 2                 4   
5153208  519140                     Minds Eyes                 3   

                  steam_id items_count            user_id  \
0        76561197970982479         277  76561197970982479   
1        76561197970982479         277  76561197970982479   
2        76561197970982479         277  76561197970982479   
3        76561197970982479         277  76561197970982479   
4        76561197970982479         277  76561197970982479   
...                    ...         ...                ...   
5153204  76561198329548331           7  76561198329548331   
5153205  76561198329548331           7  76561198329548331   
5153206  76561198329548331           7  76561198329548331   
5153207  76561198329548331           7  76561198329548331   
5153208  76561198329548331           7  76561198329548331   

                                                  user_url  
0        http://steamcommunity.com/profiles/76561197970...  
1        http://steamcommunity.com/profiles/76561197970...  
2        http://steamcommunity.com/profiles/76561197970...  
3        http://steamcommunity.com/profiles/76561197970...  
4        http://steamcommunity.com/profiles/76561197970...  
...                                                    ...  
5153204  http://steamcommunity.com/profiles/76561198329...  
5153205  http://steamcommunity.com/profiles/76561198329...  
5153206  http://steamcommunity.com/profiles/76561198329...  
5153207  http://steamcommunity.com/profiles/76561198329...  
5153208  http://steamcommunity.com/profiles/76561198329...  

[5094092 rows x 7 columns]

Se revisa los tipos de datos de cada columna, a su vez se verifica la existencia de nulos en cada una de ellas.

In [10]:
def obtener_informacion_columnas(df):
    mi_dict = {"nombre_campo": [], "tipo_datos": [], "no_nulos_%": [], "nulos_%": [], "nulos": []}

    for columna in df.columns:
        porcentaje_no_nulos = (df[columna].count() / len(df)) * 100
        mi_dict["nombre_campo"].append(columna)
        mi_dict["tipo_datos"].append(df[columna].dtype)
        mi_dict["no_nulos_%"].append(round(porcentaje_no_nulos, 2))
        mi_dict["nulos_%"].append(round(100 - porcentaje_no_nulos, 2))
        mi_dict["nulos"].append(df[columna].isnull().sum())

    df_info = pd.DataFrame(mi_dict)
    return df_info
informacion_columnas = obtener_informacion_columnas(df_user_items)
print(informacion_columnas)

       nombre_campo tipo_datos  no_nulos_%  nulos_%  nulos
0           item_id     object       100.0      0.0      0
1         item_name     object       100.0      0.0      0
2  playtime_forever      int64       100.0      0.0      0
3          steam_id     object       100.0      0.0      0
4       items_count     object       100.0      0.0      0
5           user_id     object       100.0      0.0      0
6          user_url     object       100.0      0.0      0


## Carga de datos

Una vez realizada la limpieza de datos, se cargan los mismos en un nuevo archivo `df_user_items.csv`

In [11]:
df_user_items.to_csv('Datos/df_user_items.csv', index=False)